### ISSUE TITLE
Pending server-side bucket replication prevents updating the file in the source bucket

### ISSUE THREAD

**Issue by dbulashevich** [minio/minio#16493]
## Expected Behavior
Source bucket remains fully operational even when target bucket is degraded.

## Current Behavior
File with the PENDING replication status can't be updated or remove via the source bucket

## Possible workaround
Server-side: Implement some kind of timeout for a replication attempt, i.e. switch the state from PENDING to FAILED after some timeout.

## Steps to Reproduce
1. Set up two minIO deployments (later referenced to as PRIMARY and BACKUP) with a single bucket with versioning enabled
2. Set up asynchronous server-side bucket replication from PRIMARY to BACKUP
3. Pause or slowdown the BACKUP deployment (we use `docker pause backup` in our test env)
4. Upload some file into the source bucket (`mc cp ./some_file.txt primary/bucket`)
5. Check the file stats. Replication status is PENDING, as BACKUP deployment is paused. (`mc stat primary/bucket/some_file.txt`)
6. Remove the file from the source bucket. Operation will stuck. (`mc rm primary/bucket/some_file.txt`)

Below is a code for reproducing the issue with docker compose.
https://github.com/dbulashevich/minio-playground/tree/main/replication_fail

## Context
The use case is an Active-Passive one-way replication. Ideally Active deployment shouldn't be affected by problems of the Passive backup replica while replication mode is asynchronous.

## Regression
No

## Your Environment
Host OS: Linux 5.15.0-58-generic x86_64
minio version RELEASE.2023-01-25T00-19-54Z (commit-id=50d58e9b2dbbfbe2f3afa2c16ae701f9c1cf03ec)
mc version RELEASE.2023-01-11T03-14-16Z (commit-id=14c2e506fa78b53fb6db88bcf87d8f6d3fb6989e)

**Comment by harshavardhana** [comment]
This requires locking to be version aware basically being able to mutate xl.meta allowing an incoming I/O on a pending replication attempt while it sits in failed workers () - // cc @kannappanr @poornas @klauspost 

> 8\. Copy the same file into the source bucket again. Operation will stuck

IT is not stuck as such but there is a top-level timeout to acquire a lock - since the object is being attempted as part of a "failed worker" routine - it is mostly a timing issue. 

If this failed worker fails then the status would be reflected as "FAILED" and it won't be locked no more. 

What is the use-case here - why are you re-uploading the same content again? what kind of application requires this? @dbulashevich

**Comment by klauspost** [comment]
There may be a long timeout on a replication attempt or internal retries that amplify a timeout. That should probably be investigated.

That said we have been considering per version read locks, but no decision has been made on that yet.

**Comment by poornas** [comment]
There are no internal retries in replication - the read lock on the object is not yet released because of ongoing replication attempt that has to fail first before write lock can be acquired for the new version.

 This is a timing issue before we can mark target as offline and avoid acquiring a read lock for replication

**Comment by dbulashevich** [comment]
Thanks for explanation about locking, @harshavardhana. 
Regarding the use-case: it doesn't need to be re-uploading the same content. I used re-uploading when investigating whether the problem is in our code using AWS C++ SDK for S3 protocol or replication parameters.
But the problem is reproduced when I try to remove a file in PENDING replication status. I've updated the step-to-reproduce in the issue description.
